### PR TITLE
Use an interface for the replica type.

### PR DESCRIPTION
**Pull request by jmalloc**
This change introduces an interface to represent replicas, instead of a raw struct.

The underlying implementation is now provided by a protocol buffers struct, via an adaptor (see `replicapb.Adaptor`.  

This allows us to have a protocol-buffers representation for replicas, which is fast to obtain (since it's the only implementation, at least currently), but without coupling all of the other packages to the protocol-buffers representation.

### CODE REVIEW

**Comment by jmalloc** [comment]
@koden-km I know you're not familiar with this code, but it was quite a big change, so I'd appreciate your usual veracity in telling me what I messed up, if you find the time :)

File: replica/replica.go [hunk 1]
```diff
@@ -1,80 +1,95 @@
...
 	//
 	// A replica is assigned an identifier when it starts. Once a replica goes
 	// offline it never comes back online with the same ID.
-	ID string
+	ID() string
 
-	// Addresses is a list of gRPC API addresses for the replica.
-	// At least one address must be present if using multiple replicas.
-	Addresses []string
+	// Addresses returns  a list of gRPC API addresses for the replica.
```
Comment (koden-km) [inline replica/replica.go]:
```suggestion
	// Addresses returns a list of gRPC API addresses for the replica.
```

File: replica/replicapb/adaptor.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+// ID returns the replica's unique identifier.
+//
+// A replica is assigned an identifier when it starts. Once a replica goes
+// offline it never comes back online with the same ID.
+func (a Adaptor) ID() string {
+	return a.Replica.Id
+}
+
+// Addresses returns  a list of gRPC API addresses for the replica.
```
Comment (koden-km) [inline replica/replicapb/adaptor.go]:
```suggestion
// Addresses returns a list of gRPC API addresses for the replica.
```

**Comment by jmalloc** [comment]
Fixed on master, thanks man :)

**Pull request merged**